src/main.rs

```diff
@@ -2027,62 +2027,78 @@ fn first_party_packages_strict<'a>(
 }
 
 fn check_audit_as_crates_io(cfg: &Config, store: &Store) -> Result<(), AuditAsErrors> {
-    let cache = Cache::acquire(cfg).map_err(|e| AuditAsErrors {
+    let mut cache = Cache::acquire(cfg).map_err(|e| AuditAsErrors {
         errors: vec![AuditAsError::CacheAcquire(e)],
     })?;
-    let mut needs_audit_as_entry = vec![];
-    let mut shouldnt_be_audit_as = vec![];
-
-    'packages: for package in first_party_packages_strict(&cfg.metadata, &store.config) {
-        let audit_policy = store
-            .config
-            .policy
-            .get(&package.name)
-            .and_then(|policy| policy.audit_as_crates_io);
-        if audit_policy == Some(false) {
-            // They've explicitly said this is first-party so we don't care about what's in the registry
-            continue;
-        }
 
-        if let Some(index_entry) = cache.query_package_from_index(&package.name) {
-            if storage::exact_version(&index_entry, &package.version).is_some() {
-                // We found a version of this package in the registry!
-                if audit_policy == None {
-                    // At this point, having no policy is an error
-                    needs_audit_as_entry.push(NeedsAuditAsError {
-                        package: package.name.clone(),
-                        version: package.version.clone(),
-                    });
+    // If we don't have a registry, we can't check audit-as-crates-io.
+    if !cache.has_registry() {
+        return Ok(());
+    }
+
+    'retry: loop {
```

Comment:
This is a bit gross, but adding this extra loop was the easiest way to make sure we re-check everything, as we'd otherwise get inconsistent results.

Reply:
Is it feasible to factor this into a helper or two rather than adding yet another jump label?

Reply:
I ended up moving the cache acquire out and writing it as a recursive pattern, which is I think slightly less gross to read.

Reply:
Much nicer, thanks.